src/main/kotlin/org/pkl/lsp/packages/dto/PklProject.kt

```diff
@@ -0,0 +1,107 @@
+import org.pkl.lsp.packages.Dependency
+import org.pkl.lsp.packages.LocalProjectDependency
+import org.pkl.lsp.packages.PackageDependency
+import org.pkl.lsp.packages.toDependency
+import org.pkl.lsp.util.URISerializer
+
+data class PklProject(val metadata: DerivedProjectMetadata, val projectDeps: ProjectDeps?) {
+
+  companion object {
+    private val json = Json { ignoreUnknownKeys = true }
+
+    fun parseMetadata(input: String): List<DerivedProjectMetadata> {
+      return json.decodeFromString(input)
+    }
+
+    private fun parseDeps(input: String): ProjectDeps {
+      return json.decodeFromString(input)
+    }
+
+    fun loadProjectDeps(file: VirtualFile): ProjectDeps {
+      val contents = file.contents()
+      return parseDeps(contents)
+    }
+
+    @Serializable
+    data class DerivedProjectMetadata(
+      @Serializable(with = URISerializer::class) val projectFileUri: URI,
+      val packageUri: PackageUri?,
+      val declaredDependencies: Map<String, PackageUri>,
+      val evaluatorSettings: EvaluatorSettings?,
+    )
+
+    @Serializable data class EvaluatorSettings(val moduleCacheDir: String? = null)
+
+    @Serializable
+    data class ProjectDeps(
+      val schemaVersion: Int,
+      val resolvedDependencies: Map<String, ResolvedDependency>,
+    ) {
+      /** Given a package URI, return the resolved dependency for it. */
+      fun getResolvedDependency(packageUri: PackageUri): ResolvedDependency? {
+        val packageUriStr = packageUri.toString()
+        return resolvedDependencies.entries.find { packageUriStr.startsWith(it.key) }?.value
+      }
+    }
+  }
+
+  val projectFile: Path by lazy { Path.of(metadata.projectFileUri) }
+
+  val projectDir: Path by lazy { projectFile.parent }
+
+  /** The dependencies declared within the PklProject file */
+  val myDependencies: Map<String, Dependency>
+    get() {
+      return metadata.declaredDependencies.entries.fold(mapOf()) { acc, (name, packageUri) ->
+        val dep = projectDeps?.getResolvedDependency(packageUri)?.asDependency ?: return@fold acc
+        acc.plus(name to dep)
+      }
+    }
+
+  fun getResolvedDependencies(context: PklProject?): Map<String, Dependency> {
+    if (context == null || context === this) return myDependencies
+    return myDependencies.mapValues { (_, dep) ->
+      // best-effort approach; if the dependency is not found in [context], show what's found in
+      // [myDependencies].
+      context.projectDeps?.getResolvedDependency(dep.packageUri)?.toDependency(context) ?: dep
+    }
+  }
+
+  private val self = this
+
+  private val ResolvedDependency.asDependency: Dependency?
+    get() =
+      when (this) {
+        is LocalDependency -> {
+          val localPath = projectDir.resolve(path)
+          if (Files.exists(localPath)) LocalProjectDependency(uri, localPath) else null
+        }
+        else -> {
+          this as RemoteDependency
+          PackageDependency(uri, self, checksums)
+        }
```

Comment:
Not sure if this is better? `ResolvedDependency` is sealed. Adding a type assertion here seems better than silently failing, because if we ever hit it, it's a programmer error.